text/0000-write-pointers.md

```diff
@@ -0,0 +1,484 @@
+    
+    if foo.a == 12 {
+        // When we panic here, we should drop all values that are initialized in the function.
+        // Nothing could have been initialized before the function because we have a &uninit T
+        panic!("Oh no, something went wrong!");
+    } 
+
+    foo.c = Bar { d = 10, e = 12.0 };
+}
+
+fn out_bar_panics(foo: &out Bar) {
+    // When we panic here we drop here we don't ever drop any value behind a &out because &out can never have a destructor, it doesn't matter
+    panic!("Oh no, something went wrong!");
+}
+
+let x: Foo;
+
+init_foo_can_panic(&uninit x);
+
+let x: Bar;
+
+out_bar_panics(&out x); // when we take a &out, we are asserting that the old value doesn't need to drop, and doesn't matter. This is fine because Bar is Copy and does not have a destructor.
+```
+
+# Drawbacks
+
+[drawbacks]: #drawbacks
+
+ - This is a significant change to the language and introduces a lot of complexity. \
+ - Partial initialization can be solved entirely through the type-system as shown [here](https://scottjmaddox.github.io/Safe-partial-initialization-in-Rust/). But this does have its problems, such as requiring an unstable feature (untagged_unions) or increased size of the uninitialized value (using enums).
+
+# Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+## `T: !Drop` for `&out T`
+
+If negative trait bounds become stable, the bounds for `&out T` will change to `T: !Drop`. But this does not seem like the correct bound, see [here](#unresolved-questions) for why.
+
+## Allow Drop types to be partially initialized
+
+Then they would only be dropped if all of their fields are initialized
+
+## Placement-new
+
+Placement new would help, with initializing large structs.
+
+## As sugar
+
+This could be implemented as sugar, where all fields of structs that are partially initialized are turned into temp-variables that are then passed through the normal pipeline.
+
+For example
+```Rust
+let x: Bar;
+x.d = 10;
+x.e = 12.0;
+```
+
+would desugar to
+
+```Rust
+let x: Bar;
+let xd = 10;
+let xe = 12.0;
+x = Bar { d: xd, e: xe };
+```
+
+But this would not be able to replace placement new as it can't handle `&uninit T` through function boundaries. Also this would not solve the problem of direct-initialization.
+
+# Prior art
+[prior-art]: #prior-art
+
+Out pointers in C++, (not exactly the same, but similar idea)
+
+`out T` in  C#
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+ - What is the correct bounds for `T` in `&out T`? conservatively, `T: Copy` works and is sound. But as [@TechnoMancer](https://internals.rust-lang.org/u/TechnoMancer) pointed out, `T: !Drop` is not the correct bound. For example, `Wrapper(Vec<String>)`, clearly cannot be overwritten safely, because `Vec<String>`, must drop do deallocate the `Vec`, but `Wrapper` itself does not implement drop. Therefore either a new trait is needed (but unwanted), or we must keep the `Copy` bound.
```

Comment:
[Failing to call destructors is safe](/rust-lang/rfcs/pull/1066); better find another word. This is mostly a technicality, however; resource leaks are still something Rust would rather avoid, and I agree not accounting for this here at all would make them too easy.

But besides this, I've been long thinking that `Drop` needs an overhaul. I'd replace it with something which looks like this:
- `trait Drop { /* ... */ }`: same interface as before
- `unsafe trait Destruct: Drop {}`: marker trait, assures that values of a given type can be disposed of by merely calling `Drop` implementations of their constituent fields; the type has no disposal logic of its own and can be safely destructured (this corresponds to what is now informally expressed as `!Drop`)
- `unsafe auto trait Forget: Destruct {}`: marker trait, assures that all `Drop` implementations of transitively constituent fields are no-ops; in other words, `mem::drop` on the type is completely equivalent to `mem::forget`.

Now, automatic deriving behaviour:
- for any given type `T`, unless the containing module provides an explicit `Drop` implementation, the compiler automatically derives a no-op `Drop` and `Destruct` for `T`;
- if all constituent fields are also `Forget`, the compiler derives `Forget` as well.

This system would remove a major motivation for introducing negative traits/bounds (which have been pointed out to be rather problematic, especially in the context of automatic impls for `!`), as it would create a way to formalise 'trivial disposal behaviour' in the type system without them; in this case, the correct bound would be `Forget`. It would also open the doors to adding true linear typing to Rust: those could simply explicitly opt-out of implementing `Drop`. In other words, lacking an `impl Drop` would no longer mean 'trivial disposal logic', but 'no automatic disposal at all'.

However, I'm not sure how backwards compatible this would be. Obviously, introducing true linear typing would incur some major changes.

Reply:
The Drop overhaul sounds nice on paper, I would still still want some form of negative trait impls, because there are some APIs to need this to be work at all. But that is beside the point. The Drop overhaul doesn't fit in with this RFC, and you could post it to internals if you want more feedback.

I did note that failing to call destructors is safe as the very next sub bullet. This is an unresolved question because I wasn't sure if we wanted to allow leaks this way.